xwiki-platform-core/xwiki-platform-office/xwiki-platform-office-importer/src/main/java/org/xwiki/officeimporter/internal/filter/HeadingInOrderedListFilter.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.officeimporter.internal.filter;
+
+import java.util.Map;
+import java.util.List;
+import java.util.ArrayList;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.xml.html.filter.AbstractHTMLFilter;
+
+
+/**
+ * Sometimes the heading becomes wrapped in an ordered list structure such as this during import:
+ * 1. 
+ * 11. 
+ * 111. 
+ * 1111. 
+ * 11111. (((
+ * ===== Heading text =====
+ * )))
+ *
+ * This filter removes the ordered list to produce the following:
+ *
+ * ===== Heading text =====
+ *
+ * The filter assumes that the ((( ))) has already been removed by the RedunancyFilter.
+ *
+ * Sometimes the importer also fails to terminate the ordered list such that the entire following section is included
+ * inside the list.  This filter also handles these cases.
+ *
+ * The algorithm is as follows:
+ *
+ * For all heading tags,
+ *
+ *   Find an ancestor chain consisting of ol or li tags, the topmost being an ol.
+ *
+ *   In the DOM tree rooted at the topmost ol, remove all ol and li tags of ol lists while keeping the contents of the
+ *   tags.
+ *
+ *   Replace the topmost ol with the preserved nodes in order in the original DOM.
```

Comment:
Would be nice if the list of office importer cleaning filters were configurable so that an advanced user can disable some filters if they produce bad results for a particular input document. As a start we could configure the list of filters in ``xwiki.properties`` but ideally it should be possible to select/deselect the filters from the office importer UI, using checkboxes.

This would allow the users to disable the ``HeadingInOrderedListFilter`` for documents where headings are put inside ordered lists intentionally.

But this is for another issue/commit.